### PR TITLE
feat(bundler): add --windows-icon, --windows-no-console, fix bun.exe's main icon

### LINKED ISSUE #10823: Single-file executables icon on Windows
### What is the problem this feature would solve?

Currently, the icon of an exe file built as a single-file executable with bun is a cute bun icon and cannot be changed.

### What is the feature you are proposing to solve the problem?

It would be great if the build command allows us to specify an icon file like this:
`bun build ./index.ts --compile --outfile hello.exe  --icon ./nikuman.ico`

### What alternatives have you considered?

The icon can be changed with Resource Hacker. However, it must be done each time you rebuild.

### LINKED ISSUE #13084: Allow disabling the console in `--compile` on windows. (`/SUBSYSTEM:WINDOWS`)
### What is the problem this feature would solve?

Windows GUI

### What is the feature you are proposing to solve the problem?

golang > `go build -ldflags="-s -w -H windowsgui"`

The terminal window will not be displayed.

### What alternatives have you considered?

_No response_

### LINKED ISSUE #14062: Latest `bun.exe` missing file metadata
### What version of Bun is running?

1.1.29+6d43b3662

### What platform is your computer?

Microsoft Windows NT 10.0.22631.0 x64

Hi, 
The newest Bun version, it doesn't have Bun icon on .exe file, unlike previous ones.

### PR DESCRIPTION
**Pull request by paperclover**
Missing tests because writing one is very non-trivial.

`--windows-icon` is currently gated to windows build hosts (no cross compilation), as the current approach uses a library that calls the Windows API. In the future, we should replace this with a portable version that just manipulates the bytes in-memory.

Fixes #10823
Fixes #13084

### CODE REVIEW

**Comment by robobun** [comment]
<div><sup>Updated 9:33 PM PT - Dec 19th, 2024</sup></div>

@dylan-conway, your commit 8116f33dd6648ee0f29dc8a2ae0cdc25ac256e3c is building: [`#8314`](https://buildkite.com/bun/bun/builds/8314)

File: src/cli.zig [hunk 1]
```diff
@@ -290,7 +290,9 @@ pub const Arguments = struct {
...
         clap.parseParam("--server-components              (EXPERIMENTAL) Enable server components") catch unreachable,
-        clap.parseParam("--env <inline|prefix*|disable>    Inline environment variables into the bundle as process.env.${name}. Defaults to 'inline'. To inline environment variables matching a prefix, use my prefix like 'FOO_PUBLIC_*'. To disable, use 'disable'. In Bun v1.2+, the default is 'disable'.") catch unreachable,
+        clap.parseParam("--env <inline|prefix*|disable>   Inline environment variables into the bundle as process.env.${name}. Defaults to 'inline'. To inline environment variables matching a prefix, use my prefix like 'FOO_PUBLIC_*'. To disable, use 'disable'. In Bun v1.2+, the default is 'disable'.") catch unreachable,
+        clap.parseParam("--windows-hide-console           When using --compile targetting Windows, prevent a Command prompt from opening alongside the executable") catch unreachable,
+        clap.parseParam("--windows-icon <STR>             When using --compile targetting Windows, assign an executable icon") catch unreachable,
```
Comment (dylan-conway) [inline src/cli.zig]:
```suggestion
        clap.parseParam("--windows-hide-console           When using --compile targeting Windows, prevent a Command prompt from opening alongside the executable") catch unreachable,
        clap.parseParam("--windows-icon <STR>             When using --compile targeting Windows, assign an executable icon") catch unreachable,
```

**Review by dylan-conway — APPROVED** [review]

**Review by Jarred-Sumner — CHANGES_REQUESTED** [review]
Does not successfully link on Windows.

> lld-link: error: codegen\windows-app-info.rc: unknown file type

**Comment by 716774** [comment]
I love bun ~

**Comment by CosmoMyzrailGorynych** [comment]
@paperdave see [resedit-cli](https://www.npmjs.com/package/resedit-cli) and their [resedit](https://www.npmjs.com/package/resedit) and [pe-library](https://www.npmjs.com/package/pe-library) npm packages for a purely JS solution for patching windows executables. I've been already using them for Neutralino, NW.JS and Bun binaries for a long time to edit and sign windows executables in a Linux CI.

`resedit-cli` has a nice programmatic API despite being primarily a CLI tool. You might want to use `resedit` directly to exclude unneeded bulk.

**Comment by CosmoMyzrailGorynych** [comment]
[Here is what I'm doing in my Buntralino](https://github.com/buntralino/buntralino-cli/blob/main/commands/build.ts#L70-L117) — I implemented icon addition and no-console before that PR.

**Comment by Nanome203** [comment]
> Does not fix Bun's own icon. But it's better than status quo.

Bun's main icon stopped appearing since v1.1.28

**Pull request merged**